conda/core/link.py

```diff
@@ -515,49 +537,93 @@ def _verify(cls, prefix_setups, prefix_action_groups):
-                        pkg_idx_tracked = pkg_idx + pkg_idx_start
-                        cls._execute_actions(pkg_idx_tracked, axngroup)
-                    # Run post-unlink scripts
-                    for pkg_idx, axngroup in enumerate(unlink_actions):
-                        if axngroup.type in ('unlink', 'link'):
-                            pkg_idx_tracked = pkg_idx + pkg_idx_start
-                            cls._execute_post_link_actions(pkg_idx_tracked, axngroup)
-
-                    pkg_idx_start = len(unlink_actions)
-                    # Execute link actions
-                    for pkg_idx, axngroup in enumerate(link_actions):
-                        pkg_idx_tracked = pkg_idx + pkg_idx_start
-                        cls._execute_actions(pkg_idx_tracked, axngroup)
-                    # Run post-link scripts
-                    for pkg_idx, axngroup in enumerate(link_actions):
-                        if axngroup.type in ('unlink', 'link'):
-                            pkg_idx_tracked = pkg_idx + pkg_idx_start
-                            cls._execute_post_link_actions(pkg_idx_tracked, axngroup)
-
-            except CondaMultiError as e:
+            exceptions = []
+            with Spinner("Executing transaction", not context.verbosity and not context.quiet,
+                         context.json):
+
+                # Execute unlink actions
+                with ThreadLimitedThreadPoolExecutor() as executor:
+                    for (group, register_group, install_side) in (
+                            (unlink_actions, "unregister", False),
+                            (link_actions, "register", True)):
+
+                        for axngroup in group:
+                            is_unlink = axngroup.type == 'unlink'
+                            target_prefix = axngroup.target_prefix
+                            prec = axngroup.pkg_data
+                            run_script(target_prefix if is_unlink else prec.extracted_package_dir,
+                                       prec,
+                                       'pre-unlink' if is_unlink else 'pre-link',
+                                       target_prefix)
+
+                        # parallel block 1:
+                        futures = (executor.submit(cls._execute_actions, axngroup)
+                                   for axngroup in group)
+                        for future in as_completed(futures):
+                            exc = future.result()
+                            if exc:
+                                log.debug('%r'.encode('utf-8'), exc.errors[0], exc_info=True)
+                                exceptions.append(exc)
+
+                        # Run post-link or post-unlink scripts and registering AFTER link/unlink,
+                        #    because they may depend on files in the prefix.  Additionally, run
+                        #    them serially, just in case order matters (hopefully not)
```

Comment:
If I recall correctly order matters. There was one package I ran into at least that the order that the post link scripts were running in was messed up and it broke the package sometimes.

Reply:
order is preserved, according to the original order in the axngroup.  I'm not certain that this is exactly the same as you used to have.